### PR TITLE
Data refactor

### PR DESCRIPTION
**Pull request by hyaluronic**

### CODE REVIEW

File: Marcelena/Domain/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        public static char UNDER_SCORE => '_';
+        public static char DASH => '-';
+
+
+
+
+
+
+
+
```
Comment (hyaluronic) [inline Marcelena/Domain/Constants.cs:57]: Delete unnecessary lines

File: Marcelena/Forms/FormMenuAddPlaces.cs [hunk 2]
```diff
@@ -37,22 +37,29 @@ private void customizeDesign()
...
         private void closeButton_Click(object sender, EventArgs e)
         {
             this.Close();
         }
 
         private void FormMenuAddPlaces_Load(object sender, EventArgs e)
         {
-            listBoxProduct.Items.Add(string.Format(stdDetails, "Product", "Price"));
+            //listViewProducts.Items.Add(string.Format("Product", "Price"));
         }
```
Comment (hyaluronic) [inline Marcelena/Forms/FormMenuAddPlaces.cs:57]: Method does nothing - delete

File: Marcelena/Forms/FormMenuAddPlaces.cs [hunk 3]
```diff
@@ -64,127 +71,86 @@ private void enterPressed(object sender, KeyEventArgs e)
...
-            {
-                result += p.ProductType.Replace(' ','_') + "-" + p.ProductPrice + " ";
+                    var concatProductStrings = listViewProducts.Items.Cast<ListViewItem>()/*;*/.Select(lvi => $"{ lvi.SubItems[0].Text} { lvi.SubItems[1].Text}");
+                    foreach (string item in concatProductStrings)
+                    {
+                        string[] productString = item.Split(' ');
+                        products.Add(new Product(productString[0], Double.Parse(productString[1])));
+                    }
+                    place.Products = products;
+                    PlacesService.writeShopToFile(place);
```
Comment (hyaluronic) [inline Marcelena/Forms/FormMenuAddPlaces.cs:113]: Too much logic for a form

File: Marcelena/Forms/FormMenuSearch.cs [hunk 4]
```diff
@@ -177,13 +140,70 @@ private void CloseChildForm()
...
         {
+            comboBoxOrderBy.Visible = true;
+            searchBox.Visible = true;
+            searchButton.Visible = true;
+            textBoxOrderBy.Visible = true;
             CloseChildForm();
             panelChildForm.Visible = false;
             listSearchResults.Visible = true;
             ButtonCloseForm.Visible = false;
         }
```
Comment (hyaluronic) [inline Marcelena/Forms/FormMenuSearch.cs:153]: Maybe a separate method to set all of these true or false

File: Marcelena/Forms/FormSelectedSearchResult.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                listBoxProduct.Items.Add(string.Format(stdDetails, p.ProductType, p.ProductPrice));
+            }
+        }
+
+        private async void customizeDesign(Shop shop)
+        {
+            listBoxProduct.Items.Add(string.Format(stdDetails, Constants.PRODUCT, Constants.PRICE));
+            listBoxProduct.Items.Add(string.Format(stdDetails, Constants.DEVIDER, Constants.EMPTY_STRING));
+            distance.Text = Constants.DISTANCE_TO + await GeocodeDistanceService.CalculateDistanceAsync("Didlaukio g. 59, Vilnius, Lithuania", shop.ShopAddress.ToString()) + Constants.KM;
+        }
```
Comment (jokubasjurk) [inline Marcelena/Forms/FormSelectedSearchResult.cs:41]: probably can change it so async method is not needed

File: Marcelena/Services/PlacesService.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            {
+                products.Add(product);
+            }
+        }*/
+        /*
+        public static void RemoveProductFromProducts(int index)
+        {
+            products.RemoveAt(index);
+        }
+        */
```
Comment (jokubasjurk) [inline Marcelena/Services/PlacesService.cs:32]: comments shouldn't be left out in master branch, implement it or delete it.

File: Marcelena/Services/GeocodeDistanceService.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    public class GeocodeDistanceService
+    {
+        public static async Task<string> CalculateDistanceAsync(string userAddress, string shopAddress)
+        {
+            BingMapsGeocoder geocoder = new BingMapsGeocoder(Constants.BING_MAPS_GEOCODER);
+            IEnumerable<BingAddress> userAddresses = await geocoder.GeocodeAsync(userAddress)/*.GetAwaiter().GetResult()*/;
+            IEnumerable<BingAddress> shopAddresses = await geocoder.GeocodeAsync(shopAddress);
+            var aCoord = new GeoCoordinate(userAddresses.First().Coordinates.Latitude, userAddresses.First().Coordinates.Longitude);
+            var bCoord = new GeoCoordinate(shopAddresses.First().Coordinates.Latitude, shopAddresses.First().Coordinates.Longitude);
+            return formatDistance(aCoord.GetDistanceTo(bCoord));
```
Comment (hyaluronic) [inline Marcelena/Services/GeocodeDistanceService.cs:20]: return a double. And lets try to unpack it here

File: Marcelena/Services/PlacesService.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            if (!place.ContainsProduct(product))
+            {
+                products.Add(product);
+            }
+        }*/
+        /*
+        public static void RemoveProductFromProducts(int index)
+        {
+            products.RemoveAt(index);
+        }
```
Comment (hyaluronic) [inline Marcelena/Services/PlacesService.cs:31]: delete

File: Marcelena/Services/Sort.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                                s.Products.Remove(p);
+                            }
+                            break;
+                        case ">":
+                            if (p.ProductPrice <= number)
+                            {
+                                s.Products.Remove(p);
+                            }
+                            break;
+                        case "==":
```
Comment (hyaluronic) [inline Marcelena/Services/Sort.cs:34]: use enum

File: Marcelena/Services/UpdatePlaceService.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+            return !ShopsList.All(s => !s.ShopName.Equals(searchedShop));
+        }
+
+        public static /* Shop */ void UpdateShopsInFile(Shop searchedShop)
+        {
+            bool found;
+            List<Shop> ShopsList = GetAllShops();
+            foreach (Shop s in ShopsList)
+            {
+                if(isSameShop(searchedShop, s))
```
Comment (jokubasjurk) [inline Marcelena/Services/UpdatePlaceService.cs:32]: You could make isSameShop a helper method inside Shop class for better reusability

File: Marcelena/Services/Validation.cs [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+namespace Marcelena.Services
+{
+    class Validation
+    {
+        public static string StreetNameValidation(string street)
+        {
+            //Regex rgx = new Regex("[A-Za-z0-9\\p{L}]+|[\\.-]+");
+            Regex rgx = new Regex("^[A-Za-z-\\s\\p{L}]+[\\s\\.-]*[0-9]+[a-z\\-\\d]{0,5}$");
+            //Better regex(?) "[A-Za-z- \\p{L}]+[\\.- ]*[0-9]+"
```
Comment (hyaluronic) [inline Marcelena/Services/Validation.cs:18]: delete commented lines

File: Marcelena/Services/UpdatePlaceService.cs [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+                            }
+                        }
+                        if(!found)
+                        {
+                            s.AddProduct(sp);
+                        }
+                    }
+                    //Shop s Delete line by index method
+                    // return shop s; new FileWriter(Directory.GetParent(Environment.CurrentDirectory).Parent.FullName + "\\database.txt", formatAddPlaceData(s));
+                    break;
```
Comment (jokubasjurk) [inline Marcelena/Services/UpdatePlaceService.cs:52]: again left out comments

File: Marcelena/Services/searchService.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            }
+            switch (orderBy)
+            {
+                case "shop name":
+                    results.Sort(new ShopNameComparer());
+                    break;
+                case "shop address":
+                    results.Sort(new ShopAddressComparer());
+                    break;
+                case "product price":
```
Comment (jokubasjurk) [inline Marcelena/Services/searchService.cs:50]: enum should be used here

File: Marcelena/Services/searchService.cs [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+        public static List<Shop> SortShops(List<Shop> results, string sortBy, string eq, double number)
+        {
+            switch (sortBy.ToLower())
+            {
+                case "price":
+                    results = Sort.PriceSort(results, eq, number);
+                    break;
+                case "distance":
+                   // Sort.DistanceSort(results, eq, number);
+                    break;
```
Comment (jokubasjurk) [inline Marcelena/Services/searchService.cs:86]: enum

File: Marcelena/Forms/FormMenuSearch.cs [hunk 15]
```diff
@@ -36,112 +37,57 @@ private void ButtonClose_Click(object sender, EventArgs e)
...
-                case "product type":
-                    foreach (Shop r in results)
-                    {
-                        r.Products.Sort(new ProductTypeComparer());
-                    }
-                    break;
-            }
-
+            SearchService.OrderShops(results, searchBox.Text, comboBoxOrderBy.Text.ToLower());
+            results = SearchService.SortShops(results, searchBox.Text, comboBoxSortByEQ.Text, double.Parse(comboBoxSortByNumber.Text));
```
Comment (hyaluronic) [inline Marcelena/Forms/FormMenuSearch.cs:69]: needs validation so if eq or number is not set, not to go to the method

File: Marcelena/Services/UpdatePlaceService.cs (hunk 12 above)
Comment (hyaluronic) [reply Marcelena/Services/UpdatePlaceService.cs:52]: That will be needed for future

**Pull request merged**